feat: Workflows list page performance improvements

sunyeongchoi opened the pull request:
<!-- markdownlint-disable MD041 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #12025

### Motivation
<!-- TODO: Say why you made your changes. -->
A performance issue occurred when merging the workflows and archived workflows pages into one place, and the logic was changed to resolve this.

### Modifications
<!-- TODO: Say what changes you made. -->
* When archived, an archived label is attached to the k8s resource, so the performance problem caused by the entire search was solved by changing the logic using the labelSelector of k8s.
* To manage the pagination values ​​of workflows and archived workflows separately, a value called paginationOptions was added to both the front and back.

### Verification
<!-- TODO: Say how you tested your changes. -->
1. created a workflows resource with archived enabled.
2. checked that the data was not displayed in duplicate.
3. checked if it operates normally when changing the number of pages and clicking the `next page` and `first page` buttons.
4. checked whether all test codes pass without any problems

tooptoop4 commented:
@sunyeongchoi are u able to include fix for https://github.com/argoproj/argo-workflows/issues/12161 too?

terrytangyuan reviewed (commented):
Thank you! Let's wait for others to review as well this time. It's Thanksgiving week and so please be patient while waiting for feedback.

sunyeongchoi commented:
> @sunyeongchoi are u able to include fix for #12161 too?

Yes. But I will take a look after this PR is completely finished.

sunyeongchoi commented:
I found a bug in the code I implemented. (Cronjob, event, and cli commands were also using a workflow list, but I realized that this had not been considered in this PR.)

As a result, the current PR is not complete and plan to add additional commit.

However, I would like to receive a review to see if the PR I implemented is in line with the direction of resolving the issue.

for example, when combining k8s workflows and archived workflows to show them on one page, the `PaginationOptions` parameter was added to the parameters, and the names were set to `WfContinue` for k8s workflows continue and `ArchivedContinue` for archived workflows continue.

I would like to receive reviews on whether you think this name is right or if there are other good names.

terrytangyuan commented:
> When archived, an archived label is attached to the k8s resource, so the performance problem caused by the entire search was solved by changing the logic using the labelSelector of k8s.

Could you elaborate why this would solve the performance problem? If the number of archived workflows is large, it seems like we would still encounter the performance issue.

sunyeongchoi commented:
> Could you elaborate why this would solve the performance problem? If the number of archived workflows is large, it seems like we would still encounter the performance issue.

Okay. A performance problem occurred because the total number of `archived workflows` was searched.

So the logic I implemented is as follows:

1. If the label "Archived" is attached to `k8s workflows`, it means that it is stored in `archived workflows`. Therefore, by adding the labelSelector syntax to `k8s workflows` `wfLabelSelector += fmt.Sprintf("%s!=%s", common.LabelKeyWorkflowArchivingStatus, "Archived")`, workflows stored in `archived workflows` are not searched in `k8s workflows`. (In other words, the possibility that `k8s workflows` and `archived workflows` may overlap was resolved through kubernetes labelSelector.)

2. If 10 pieces of data are needed on one page, if 6 `k8s workflows` were searched in step 1, 10-6 = 4, that is, there is a possibility of searching 4 more data, so at this time, only search up to 4 `archived workflows`.
```go
if int64(len(mergedWfs)) < limit {
  listOption.Continue = archivedContinue
  listOption.Limit = limit - int64(len(mergedWfs))
  listOption.LabelSelector = labelSelector
  archivedWfList, err = s.wfArchiveServer.ListArchivedWorkflows(ctx, &workflowarchivepkg.ListArchivedWorkflowsRequest{
    ListOptions: listOption,
    NamePrefix:  "",
    Namespace:   req.Namespace,
  })
}
```

3. As described in number 1, deduplication was solved through Kubernetes' labelSelector, and the page limit was solved as described in number 2, so if `k8s workflows` and `archived workflows` manage only the continue value for each pagination separately, the performance problem will be solved because `k8s workflows` and `archived workflows` use their own pagination logic.
```go
res := &wfv1.WorkflowList{ListMeta: metav1.ListMeta{ResourceVersion: wfList.ResourceVersion}, Items: mergedWfs, PaginationOptions: wfv1.WorkflowPaginationOptions{
  WfContinue:       common.CheckNilString(wfList.ListMeta.Continue),
  ArchivedContinue: common.CheckNilString(archivedWfList.ListMeta.Continue),
}}
```

agilgur5 commented on ui/src/models/workflows.ts:
```diff
@@ -776,6 +776,10 @@ export interface Condition {
+export interface WorkflowsPaginationOptions {
+    wfContinue: string;
+    archivedContinue: string;
+}
```
```suggestion
}

```

missing a new line as is convention in this file between all the types

agilgur5 commented on ui/src/app/workflows/pagination.ts:
```diff
@@ -0,0 +1,7 @@
+export interface WorkflowsPagination {
```
I imagine this could just live next `WorkflowsPaginationOptions` in `models/workflows.ts`, no?

sunyeongchoi replied:
Oh. that is more natural! I'll reflect this opinion. Thank you :)

agilgur5 commented on ui/src/app/shared/services/workflows-service.ts:
```diff
@@ -36,6 +38,35 @@ export const WorkflowsService = {
+            'items.status.estimatedDuration',
+            'items.status.progress',
+            'items.spec.suspend',
+            'paginationOptions.wfContinue',
+            'paginationOptions.archivedContinue'
```
Oh it looks like you added `paginationOptions` to the `WorkflowsList` struct in `workflow_types.go` [above](https://github.com/argoproj/argo-workflows/pull/12237/files#diff-94d0f1284fafd0a1cb05dc2309b4be043b6fb01d66626ab1867d40c84b573ec2R251), nevermind then

agilgur5 commented on pkg/apiclient/workflow/workflow.proto:
```diff
@@ -147,6 +148,10 @@ service WorkflowService {
     option (google.api.http).get = "/api/v1/workflows/{namespace}/{name}";
   }
 
+  rpc ListK8SAndArchivedWorkflows(WorkflowListRequest) returns (github.com.argoproj.argo_workflows.v3.pkg.apis.workflow.v1alpha1.WorkflowList) {
+    option (google.api.http).get = "/api/v1/k8s-archived-workflows/{namespace}";
+  }
+
```
So I was thinking that this would work more roughly within the constraints of the v3.5 API, as in, implementing it as part of the `workflows` API method just as optional parameters. Would that not work?

sunyeongchoi replied:
I also kept thinking about that part.
If so, I will reflect this by branching to parameters within `ListWorkflows` and separating them.

Thank you for your feedback!

agilgur5 reviewed (commented):
I only looked through this pretty quickly, so I still need to spend some more time reviewing this, but left a few comments.

Generically speaking, having two cursors (one for archived and one for in-cluster) was my idea to fix this as well, so I think I'm roughly good with this approach, but I need to look through some of the specifics a bit more. The API method, the deduplication approach, and how the cursors are stored in particular stood out.

I think this is roughly in the right direction, thanks for all the work and testing to get this functional!

tczhao commented:
hi @sunyeongchoi , could you fix the DCO signature

agilgur5 commented:
> hi @sunyeongchoi , could you fix the DCO signature

This still needs more work in any case, per above review.

Also per my [last comment](https://github.com/argoproj/argo-workflows/issues/12025#issuecomment-1881912722) on the issue, Sun Yeong will be busy a few more days at least

sunyeongchoi commented:
https://github.com/argoproj/argo-workflows/issues/12025#issuecomment-1949330952

I believe the concept of the issue has completely changed, so if everyone agrees, I will close this PR.

terrytangyuan commented:
Thank you so much for your efforts so far!
